### ISSUE TITLE
Lame leaper that has to be blocked on more than 1 square?

### ISSUE THREAD

**Issue by Yowwowwow** [fairy-stockfish/Fairy-Stockfish#642]
I wanna implement a piece but am not sure how to. This piece moves like a king, but when moving diagonally, it can be blocked from a square if it cannot reach that square in 2 wazir moves (1st wazir move can't be a capture). For example, it cannot move to its top-right square if its right square and top square are both occupied.
How can I implement this piece? More generally, can I implement a lame leaper that has to be blocked on more than 1 square? E.g. a knight that can move from (0, 0) to (1, 2) only if at least 1 square of (0, 1) and (1, 1) is empty.

**Comment by HGMuller** [comment]
Multi-path pieces can be defined by simply equiping them with all the moves that lead to the same square. For example, the mentioned lame Knight (known as Moo) has both the moves of the Mao (blockable on the (0,1) square) and the Moa (blockable on (1,1)). But FSF does not support the Moa.

The mentioned double-Wazir piece could be implemented cheaply as a side effect of the Mao attacks. By also including F squares that border on an empty W square in the attacks table for the Mao, the same magic ride could be used for move generation. It would just use a different pseudoMoves mask, selecting the F squares from the attacks rather than the N squares. In fact the lame Dabbaba could be included in the mix too. So there would be a unified W-blockers magic (using a mask of only 4 squares), which in its attacks bitboards would contain all squares next to an empty W square. Depending on the piece you want you would either apply F, N or D pseudoMoves, to get the mentioned 'Fat-Pawn capture' move, the Mao or the Lame Dabbaba.

In theory there could be a risk in this strategy, because the pseudoMoves are recored for the entire piece, and not per move component. So if a 'combination magic' table would also contain moves to squares we did not intend to generate the lame move to, if that square happened to be reachable through another move of the piece, it would also activate the unintended lame move. For F squares this is not really conceivable; other moves to it (in any case amongst the moves supported by FSF) would all reach it through an unblockable step, as it is an adjacent square. The D move is more tricky, as in theory lame moves over the F squares could also reach it. So there would be no way to indicate we want the D squares only to be reached through, say, the adjacent W square, and not the F squares if we needed to include the latter for some other reason. (E.g. because the piece also moves as a Moa.)

Of course it would be nice to have a similar universal magic for F-blocks (so we could do the Moa).

Note the Lame Dabbaba can be generated from Rook magics as well, but albeit slightly less efficient, because of the RookV-RookH split. So there is no strict need to include it in the W_blockers magic.

**Comment by kikoarweyo** [comment]
So, there is not alternative solution for the moa movement?  I wanted to make a compound moa+ferz

**Comment by HGMuller** [comment]
I made a version of FSF that supports the bent riders Griffon and Manticore, and the Moa is in fact a limited range Griffon. I am not sure whether specifying a range on the Griffon's move notation (yafsF) would work as intended, though. I never really payed attention to that. The official notation for a Moa would by afsF anyway; not yafsF2. So I really should check whether that works correctly.

It would also not be very hard to add the Moa move as a standard magic table like that of the Horse. The tables for such pieces are quite small. This would be much more efficient speedwise: FSF would calculate the attack set with a single lookup. The Griffon has so many moves that it requires the magics to be split into four (which could fortunately use the Rook tables). The Moa is so rarely used that investing much code into making it a little faster to generate its moves might not be worth it, though. 

[Edit] I checked my source code, and it turned out that only two-leg moves of the type yafX and yafsX (ski sliders or bent sliders) with X = W or F were recognized. Without the range-toggling y the move was entirely ignored. It is a trivial change to also recognize also afX and afsX: just limit the range to two steps if the y is missing, rather than setting it to infinite (when it is present). Only the Moa (afsF) would benefit from that, as the Horse and lame Elephant and Dababbah are already implemented as seperate pieces.
